### DISCUSSION TITLE
having trouble with binding a button when it is in a itemcontrol

### DISCUSSION THREAD

**Discussion by sam223221** [AvaloniaUI/Avalonia#15492] [Q&A] [answered]
im trying to make a navigation bar, where i can change the content of the bar dynamicly but im having trouble with binding my button outside of the itemsSource ive tried using ChatGPT but he wasent to any help! this is how my navgation bar looks like
```
					<ItemsControl ItemsSource="{Binding Buttons}">
						<ItemsControl.ItemTemplate>
							<DataTemplate>
								<Button Width="55"
										Height="55"
										Margin="0,30,0,0"
										Command="{Binding  $parent[ItemsControl].DataContext.NavigationButtonPressedCommand}"
										CommandParameter="{Binding ButtonName}"
										>
									<Image Opacity="1"
										   IsVisible="True"
										   Source="{Binding ImageSource}"
										   Width="40" Height="40"/>
								</Button>
							</DataTemplate>
						</ItemsControl.ItemTemplate>
					</ItemsControl>
```
and i have a function to sort where the viewer needs to go after the button press but i can execute it
any advice?

**Comment by thevortexcloud** [comment]
>  ive tried using ChatGPT 

ChatGPT probably wont help much in general with Avalonia. There are not enough Avalonia specific examples and it looks too similar to most other XAML frameworks. Both things make most LLMs useless.


> im trying to make a navigation bar, where i can change the content of the bar dynamicly but im having trouble with binding my button outside of the itemsSource

To clarify: You are trying to bind to a specific item in the items collection? You can fairly easily do that by just declaring another property in your view model that points to that item. I don't think you can do it entirely in XAML anyway without a custom markup extension.

**Reply by sam223221** [reply]
im creating all the buttons in my viewModel, and i want each button when clicked to execute one method in the viewModel and give in a command parameter which is different to each button(on creation) and in the method i want to change the contentControl in my window 😕 but i cant figure it out how to do it!

**Reply by thevortexcloud** [reply]
Can you show me your view model code? It would help me understand where you are currently stuck.

**Comment by workgroupengineering** [comment]
See [this](https://github.com/AvaloniaUI/Avalonia.Labs/blob/main/samples/Avalonia.Labs.Catalog/Views/MainView.axaml) Sample.

**Reply by xiejiajian** [reply]
This is the solution!!!!

**Comment by sam223221** [comment, marked as answer]
i have figured it out now! thanks for your suggestions,
i made the first half of a reactivecommand in my Model so i can add it to my command though binding and when i create the button in my main VM i create the rest! here are the snippets :)

VM:
`        private void AdminNavigationBar()
        {
            NavigationButtons = new ObservableCollection<MyButtonModel>
            {
                new MyButtonModel("User", ReactiveCommand.Create<string>(AdminNavigationExe)),
                new MyButtonModel("Admin", ReactiveCommand.Create<string>(AdminNavigationExe)),
            };
        }
`

Model:
`    public class MyButtonModel 
    {

        public ReactiveCommand<string, Unit> example { get; }
        public string Name { get; set; }
        public ICommand Command { get; set; }

        public MyButtonModel(string name, ReactiveCommand<string, Unit> example)
        {
            Name = name;
            this.example = example;
        }
    }
`
and in axaml:
`				<StackPanel>
				
					<ItemsControl ItemsSource="{Binding NavigationButtons}">
						<ItemsControl.ItemTemplate>
							<DataTemplate>
								<Button Width="55"
										Height="55"
										Margin="0,10,0,0"
										Content ="{Binding Name}"
										Command="{Binding example}"
										CommandParameter="User"
										>
								</Button>
								
							</DataTemplate>
						</ItemsControl.ItemTemplate>
					</ItemsControl>

				</StackPanel>
`

here is sort of diagram of what i understand happens
![image](https://github.com/AvaloniaUI/Avalonia/assets/58514851/e08046bd-6b95-47b4-8781-543001274f7e)

**Reply by thevortexcloud** [reply]
FYI if all the items in the collection are using the same command functionality, you can just bind them all to the same command instance. You can also bind them to parent view model too using `RelativeSource`/one of the other binding ways that lets you bind to another context. You only need separate instances if you need to disable/enable the commands individually.

On another note, it actually looks like you are trying to implement a custom router. Seeing as you are using ReactiveUI already, you can just use the one that comes with it.

https://docs.avaloniaui.net/docs/concepts/reactiveui/routing

There are lots of other routers available too (EG ones that work like a website URL). But you will need to find them elsewhere.

**Reply by sam223221** [reply]
i tried using RelativeSource but i couldn't make it work it always gave me an error

**Reply by sam223221** [reply]
i tried so many ways using `$parent` and `RelativeSource` among things, this solution was the only way i could make it work :/

**Reply by thevortexcloud** [reply]
The thing about relative source is it finds controls, not view models. As such you need to bind to the datacontext of the control you are looking for. Avalonia also now uses compiled bindings by default, this is normally a good thing but it can cause errors if you don't know about it. Mainly when binding to the datacontext of a control, which will be a type of `object` rather than your view model type. As such you need to actually cast the datacontext to your view model type in your xaml to be able to bind to a specific property on it. Alternatively you can just use a `ReflectionBinding` explicitly which does not have that problem (but it does have other problems).

This is fairly different from WPF which only has Reflection bindings, which will let you get away with far more things.

There are examples on the docs site on how to do this for Avalonia:

https://docs.avaloniaui.net/docs/basics/data/data-binding/compiled-bindings
